ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -221,13 +222,150 @@ public ReplicaSyncUpManager getReplicaSyncUpManager() {
+      }
+    } else {
+      // add replica onto a brand new disk
+      Map<String, String> diskInfoToAdd = new HashMap<>();
+      diskInfoToAdd.put(ClusterMapUtils.DISK_CAPACITY_STR,
+          Long.toString(replicaId.getDiskId().getRawCapacityInBytes()));
+      diskInfoToAdd.put(ClusterMapUtils.DISK_STATE, ClusterMapUtils.AVAILABLE_STR);
+      diskInfoToAdd.put(ClusterMapUtils.REPLICAS_STR, newReplicaInfo);
+      mountPathToDiskInfos.put(replicaId.getMountPath(), diskInfoToAdd);
+      newReplicaInfoAdded = true;
+    }
+    if (newReplicaInfoAdded) {
+      // we update InstanceConfig only when new replica info is added (skip updating if replica is already present)
+      instanceConfig.getRecord().setMapFields(mountPathToDiskInfos);
+      logger.info("Updating config: {} in Helix by adding partition {}", instanceConfig, partitionName);
+      additionResult = helixAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+    }
+    return additionResult;
+  }
+
+  /**
+   * Remove old/existing replica info from {@link InstanceConfig} that associates with current data node.
+   * @param replicaId the {@link ReplicaId} whose info should be removed.
+   * @param instanceConfig {@link InstanceConfig} to update.
+   * @return {@code true} replica info is successfully removed. {@code false} otherwise.
+   */
+  private boolean removeOldReplicaInfo(ReplicaId replicaId, InstanceConfig instanceConfig) {
+    boolean removalResult = true;
+    boolean replicaFound = false;
+    String partitionName = replicaId.getPartitionId().toPathString();
+    Map<String, Map<String, String>> mountPathToDiskInfos = instanceConfig.getRecord().getMapFields();
+    Map<String, String> diskInfo = mountPathToDiskInfos.get(replicaId.getMountPath());
+    if (diskInfo != null) {
+      String replicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+      if (!replicasStr.isEmpty()) {
+        List<String> replicaInfos =
+            new ArrayList<>(Arrays.asList(replicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR)));
+        // if any element is removed, that means old replica is found in replicasStr.
+        replicaFound = replicaInfos.removeIf(
+            info -> (info.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR)[0]).equals(partitionName));
+
+        // We update InstanceConfig only when replica is found in current instanceConfig. (This is to avoid unnecessary
+        // notification traffic due to InstanceConfig change)
+        if (replicaFound) {
+          StringBuilder newReplicasStrBuilder = new StringBuilder();
+          // note that old replica info has been removed from "replicaInfos"
+          for (String replicaInfo : replicaInfos) {
+            newReplicasStrBuilder.append(replicaInfo).append(ClusterMapUtils.REPLICAS_DELIM_STR);
+          }
+          // update diskInfo and MountPathToDisk map
+          diskInfo.put(ClusterMapUtils.REPLICAS_STR, newReplicasStrBuilder.toString());
```

Comment:
If there are no more replicas on this disk (mount path) after removal, should we remove the disk from mountPathToDiskInfos?

Reply:
A fair point. We temporarily don't remove the empty disk because it might be reused for holding other new replicas. If the node will be decommissioned, we directly remove InstanceConfig associated with this node in Helix.